src/coreclr/jit/morph.cpp

```diff
@@ -14106,13 +14107,22 @@ void Compiler::fgMergeBlockReturn(BasicBlock* block)
 
             // GT_RETURN must have non-null operand as the method is returning the value assigned to
             // genReturnLocal
-            noway_assert(ret->OperGet() == GT_RETURN);
-            noway_assert(ret->gtGetOp1() != nullptr);
+            GenTree* retVal;
+            if (ret->OperIs(GT_SWIFT_ERROR_RET))
+            {
+                retVal = ret->gtGetOp2();
+            }
+            else
+            {
+                noway_assert(ret->OperIs(GT_RETURN));
+                retVal = ret->gtGetOp1();
+            }
```

Comment:
```suggestion
            GenTree* retVal = ret->GetReturnedValue();
```

It would be good to assert that the error operand is the expected one since you rely on that in `CreateReturnBB`. Essentially there is a subtle invariant here that no transformations return between import and return merging.

It would be more pay-for-play to actually wait with introducing the `GT_SWIFT_ERROR_RET` until after return merging, by adding a phase that transforms `GT_RETURN` to `GT_SWIFT_ERROR_RET` only for ``CallConvSwift` functions. That would get rid of the check in `impLoadArg` and avoid the need to change return merging entirely.

Reply:
I like the idea of doing the transformation in one place; let me give this a try.

> That would get rid of the check in impLoadArg and avoid the need to change the early part of return merging entirely.

I'm a bit confused about the `impLoadArg` part -- we still need to convert usages of the `SwiftError*` parameter into references to the pseudolocal, right? Do you mean we can get rid of the `GT_SWIFT_ERROR_RET` transformation in `impReturnInstruction`?

Reply:
> I'm a bit confused about the `impLoadArg` part -- we still need to convert usages of the `SwiftError*` parameter into references to the pseudolocal, right?

I think this replacement could be done by the new phase as well. It introduces a full IR traversal for `CallConvSwift` methods. I think that's a better trade off than doing the check in `impLoadArg`, for everyone.

> Do you mean we can get rid of the `GT_SWIFT_ERROR_RET` transformation in `impReturnInstruction`?

Right, that would be done in the new phase too.

Reply:
I see, thanks!

> It introduces a full IR traversal for CallConvSwift methods. I think that's a better trade off than doing the check in impLoadArg, for everyone.

Agreed.

Reply:
I tried out this phase approach, and if we wait to run it after global morph, we're too late to do struct promotion of the `SwiftError` pseudolocal, and the presence of `GT_LCL_FLD` usages of `SwiftError::Value` (instead of `GT_LCL_VAR`s of the promoted `Value` local) seems to trigger asserts during lowering since we're allowing the pseudolocal to be enregistered. I'm not sure if this approach is desirable...